### ISSUE TITLE
procutils.which fails with a KeyError if PATH is not in os.environ

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2313]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#2313|
|Type|defect|
|Created|2006-12-20 19:09:45Z|

It should just use os.environ.get, like it does on the line immediately before it.

<details><summary>Searchable metadata</summary>

```
trac-id__2313 2313
type__defect defect
reporter__dreid dreid
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1166641785000000 1166641785000000
changetime__1167068633000000 1167068633000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Fixed in which-without-path-2313

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_3"></a>|@jerub commented|
|-|-|

This fails with
```
trial --random 4 twisted.test.test_process.UtilTestCase
```
because the deleted key from os.environ is not restored after the test, only in tearDownClass. Please update the test so that it doesn't depend on the order of execution of tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jerub|
|-|-|

Thanks :)

I replaced setUpClass and tearDownClass with setUp and tearDown and made them use unique directories for their test files as well.  I also cleaned up the setup code and removed the code which deleted everything afterwards, since there's no good reason to do that, and at least one reason not to (easier to investigate failures afterwards if you haven't destroyed your test state).

Ready for another look in the same branch.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_5"></a>|@jerub set owner to @exarkun|
|-|-|

Okay, awesome. Thanks for cleaning that up. It's much cleaner and clearer.

One more change I'd like, and that's to remove the 'import shutil' from twisted/test/test_process.py, as you've removed the call to shutil.rmtree, but not removed the import now that shutil is no longer used.

After you'd removed that line, merge this branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [19083]) merge which-without-path-2313

Author: exarkun
Reviewer: jerub
Fixes [#2313](https://github.com/twisted/twisted/issues/2313)

In twisted.python.procutils.which, handle the case where the PATH environment
variable is not set.  Also add test coverage for this case and simplify the
existing which tests by removing their shared state.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
